moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp

```diff
@@ -0,0 +1,1012 @@
+}
+
+double Trajectory::getVelocityMaxPathVelocityDeriv(double pathPos)
+{
+  const Eigen::VectorXd tangent = path.getTangent(pathPos);
+  double maxPathVelocity = std::numeric_limits<double>::max();
+  unsigned int activeConstraint;
+  for (unsigned int i = 0; i < n; i++)
+  {
+    const double thisMaxPathVelocity = maxVelocity[i] / std::abs(tangent[i]);
+    if (thisMaxPathVelocity < maxPathVelocity)
+    {
+      maxPathVelocity = thisMaxPathVelocity;
+      activeConstraint = i;
+    }
+  }
+  return -(maxVelocity[activeConstraint] * path.getCurvature(pathPos)[activeConstraint]) /
+         (tangent[activeConstraint] * std::abs(tangent[activeConstraint]));
+}
+
+bool Trajectory::isValid() const
+{
+  return valid;
+}
+
+double Trajectory::getDuration() const
+{
+  return trajectory.back().time;
+}
+
+std::list<Trajectory::TrajectoryStep>::const_iterator Trajectory::getTrajectorySegment(double time) const
+{
+  if (time >= trajectory.back().time)
+  {
+    std::list<TrajectoryStep>::const_iterator last = trajectory.end();
+    last--;
+    return last;
+  }
+  else
+  {
+    if (time < cachedTime)
+    {
+      cachedTrajectorySegment = trajectory.begin();
+    }
+    while (time >= cachedTrajectorySegment->time)
+    {
+      cachedTrajectorySegment++;
+    }
+    cachedTime = time;
+    return cachedTrajectorySegment;
+  }
+}
+
+Eigen::VectorXd Trajectory::getPosition(double time) const
+{
+  std::list<TrajectoryStep>::const_iterator it = getTrajectorySegment(time);
+  std::list<TrajectoryStep>::const_iterator previous = it;
+  previous--;
+
+  double timeStep = it->time - previous->time;
+  const double acceleration =
+      2.0 * (it->pathPos - previous->pathPos - timeStep * previous->pathVel) / (timeStep * timeStep);
+
+  timeStep = time - previous->time;
+  const double pathPos = previous->pathPos + timeStep * previous->pathVel + 0.5 * timeStep * timeStep * acceleration;
+
+  return path.getConfig(pathPos);
+}
+
+Eigen::VectorXd Trajectory::getVelocity(double time) const
+{
+  std::list<TrajectoryStep>::const_iterator it = getTrajectorySegment(time);
+  std::list<TrajectoryStep>::const_iterator previous = it;
+  previous--;
+
+  double timeStep = it->time - previous->time;
+  const double acceleration =
+      2.0 * (it->pathPos - previous->pathPos - timeStep * previous->pathVel) / (timeStep * timeStep);
+
+  timeStep = time - previous->time;
+  const double pathPos = previous->pathPos + timeStep * previous->pathVel + 0.5 * timeStep * timeStep * acceleration;
+  const double pathVel = previous->pathVel + timeStep * acceleration;
+
+  return path.getTangent(pathPos) * pathVel;
+}
+
+namespace time_optimal_trajectory_generation
+{
+
+bool computeTimeStamps(robot_trajectory::RobotTrajectory& trajectory,
+                       const double max_velocity_scaling_factor,
+                       const double max_acceleration_scaling_factor)
+{
+  double tolerance = 0.01;  // TODO
+  double resample = 0.1;  // TODO
```

Comment:
address this :)

Reply:
Did you have a preference of the question I mentioned, regarding computeTimeStamps signature?

Reply:
No preference